klayout/drc/rule_decks/main.drc

```diff
@@ -0,0 +1,1072 @@
+
+    if METAL_LEVEL != "3LM"
+      connect(metal3,  via3)
+      connect(via3,    metal4)
+      
+      if METAL_LEVEL != "4LM"
+        connect(metal4,  via4)
+        connect(via4,    metal5)
+
+        if METAL_LEVEL != "5LM"
+          connect(metal5,  via5)
+          connect(via5,    metaltop)
+        end
+      end
+    end
+  end
+
+end #CONNECTIVITY_RULES
+
+#================================================
+#------------ PRE-DEFINED FUNCTIONS -------------
+#================================================
+
+def conn_space(layer,conn_val,not_conn_val, mode)
+  if conn_val > not_conn_val
+    raise "ERROR : Wrong connectivity implementation"
+  end
+  connected_output = layer.space(conn_val.um, mode).polygons(0.001)
+  unconnected_errors_unfiltered = layer.space(not_conn_val.um, mode)
+  singularity_errors = layer.space(0.001.um)
+  # Filter out the errors arising from the same net
+  unconnected_errors = DRC::DRCLayer::new(self, RBA::EdgePairs::new)
+  unconnected_errors_unfiltered.data.each do |ep|
+    net1 = l2n_data.probe_net(layer.data, ep.first.p1)
+    net2 = l2n_data.probe_net(layer.data, ep.second.p1)
+    if !net1 || !net2
+      puts "Should not happen ..."
+    elsif net1.circuit != net2.circuit || net1.cluster_id != net2.cluster_id
+      # unconnected
+      unconnected_errors.data.insert(ep)
+    end
+  end
+  unconnected_output = unconnected_errors.polygons.or(singularity_errors.polygons(0.001))
+  return connected_output, unconnected_output
+end
+
+def conn_separation(layer1, layer2, conn_val,not_conn_val, mode)
+  if conn_val > not_conn_val
+    raise "ERROR : Wrong connectivity implementation"
+  end
+  connected_output = layer1.separation(layer2, conn_val.um, mode).polygons(0.001)
+  unconnected_errors_unfiltered = layer1.separation(layer2, not_conn_val.um, mode)
+  # Filter out the errors arising from the same net
+  unconnected_errors = DRC::DRCLayer::new(self, RBA::EdgePairs::new)
+  unconnected_errors_unfiltered.data.each do |ep|
+    net1 = l2n_data.probe_net(layer1.data, ep.first.p1)
+    net2 = l2n_data.probe_net(layer2.data, ep.second.p1)
+    if !net1 || !net2
+      puts "Should not happen ..."
+    elsif net1.circuit != net2.circuit || net1.cluster_id != net2.cluster_id
+      # unconnected
+      unconnected_errors.data.insert(ep)
+    end
+  end
+  unconnected_output = unconnected_errors.polygons(0.001)
+  return connected_output, unconnected_output
+end
+
+# === IMPLICIT EXTRACTION ===
+if CONNECTIVITY_RULES
+  logger.info("Connectivity rules enabled, Netlist object will be generated.")
+  netlist
+end #CONNECTIVITY_RULES
+
+# === LAYOUT EXTENT ===
+CHIP = extent.sized(0.0)
+logger.info("Total area of the design is #{CHIP.area()} um^2.")
+
+#================================================
+#----------------- MAIN RUNSET ------------------
+#================================================
+
+logger.info("Starting GF180MCU DRC rules.")
+if FEOL
+  logger.info("Running all FEOL rules")
```

Comment:
are those placeholders? (maybe we should add an "UNIMPLEMENTED" warning)

Reply:
@proppy Those are just logging for user to understand what will happen. All rules are implemented in their appropriate table file.

Reply:
Yep, but we don't have all check ready for those yet, do we? (only dualgate)

Reply:
@proppy This rule deck is entirely complete, I'm only pushing one table at at time for your review. This rule deck was used for MPW-0 tapeout. And BTW, I have all the other tables ready as PRs as well now, please refer to:

https://github.com/efabless/globalfoundries-pdk-libs-gf180mcu_fd_pv/pulls

Reply:
should we run/include `dualgate` from here?

Reply:
@proppy I tried that using `require` in ruby doesn't work well in klayout for some reason. The best way to do this is to include all the rule deck here as you mentioned and to avoid the need to concatenate files in python. But I tried several solutions non worked with klayout. The solutions work outside of klayout on a regular ruby script.

References:
- https://stackoverflow.com/questions/3672586/what-is-the-difference-between-require-relative-and-require-in-ruby
 - https://medium.com/@connorstack/understanding-ruby-load-require-gems-bundler-and-rails-autoloading-from-the-bottom-up-3b422902ca0

Reply:
Opened an issue #16